File: OUROBOROS_DNA_MUTATION_ENABLED/5D-Suite/modules/dna_reconstruction.py
import hashlib
import numpy as np

class DNAReconstructor:
    def __init__(self):
        self.genbank = {}

    def extract_dna(self, seed):
        h = hashlib.sha256(seed.encode()).hexdigest()
        return [int(h[i:i+2], 16)/255.0 for i in range(0, 100, 2)]

    def store(self, zelle_id, seed):
        self.genbank[zelle_id] = self.extract_dna(seed)

    def compare(self, id1, id2):
        d1 = self.genbank.get(id1)
        d2 = self.genbank.get(id2)
        if not d1 or not d2:
            return 0.0
        sim = 1.0 - np.linalg.norm(np.array(d1) - np.array(d2)) / np.sqrt(len(d1))
        return max(0.0, min(sim, 1.0))

    def closest_match(self, reference_id):
        reference = self.genbank.get(reference_id)
        if not reference:
            return None, 0.0
        best, score = None, 0.0
        for other_id, dna in self.genbank.items():
            if other_id == reference_id:
                continue
            sim = self.compare(reference_id, other_id)
            if sim > score:
                best, score = other_id, sim
        return best, score

File: crates/phosphoros-cli/src/visual/suite5d/dna_reconstruction.rs
use std::collections::HashMap;

use sha2::{Digest, Sha256};

fn extract_dna(seed: &str) -> Vec<f64> {
    let mut hasher = Sha256::new();
    hasher.update(seed.as_bytes());
    hasher
        .finalize()
        .iter()
        .take(50)
        .map(|byte| *byte as f64 / 255.0)
        .collect()
}

#[derive(Debug, Default, Clone)]
pub struct DNAReconstructor {
    genbank: HashMap<String, Vec<f64>>,
}

impl DNAReconstructor {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn store(&mut self, zelle_id: impl Into<String>, seed: &str) {
        let dna = extract_dna(seed);
        self.genbank.insert(zelle_id.into(), dna);
    }

    pub fn compare(&self, id1: &str, id2: &str) -> f64 {
        let d1 = match self.genbank.get(id1) {
            Some(dna) => dna,
            None => return 0.0,
        };
        let d2 = match self.genbank.get(id2) {
            Some(dna) => dna,
            None => return 0.0,
        };
        let len = d1.len().min(d2.len());
        if len == 0 {
            return 0.0;
        }
        let mut sum = 0.0;
        for i in 0..len {
            let diff = d1[i] - d2[i];
            sum += diff * diff;
        }
        let norm = (sum / len as f64).sqrt();
        (1.0 - norm).clamp(0.0, 1.0)
    }

    pub fn closest_match(&self, reference_id: &str) -> Option<(String, f64)> {
        if !self.genbank.contains_key(reference_id) {
            return None;
        }
        let mut best: Option<(String, f64)> = None;
        for (other_id, _) in self
            .genbank
            .iter()
            .filter(|(id, _)| id.as_str() != reference_id)
        {
            let score = self.compare(reference_id, other_id);
            if best.as_ref().map(|(_, s)| score > *s).unwrap_or(true) {
                best = Some((other_id.clone(), score));
            }
        }
        best
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dna_comparison_returns_score() {
        let mut recon = DNAReconstructor::new();
        recon.store("cell_a", "alpha_seed");
        recon.store("cell_b", "beta_seed");
        recon.store("cell_c", "alpha_seed");
        let score = recon.compare("cell_a", "cell_c");
        assert!(score > 0.9);
        let best = recon.closest_match("cell_a").unwrap();
        assert_eq!(best.0, "cell_c");
    }
}
